### PR TITLE
[CI:BUILD] Packit: run cockpit-podman tests in PRs

### PR DESCRIPTION
**Pull request by martinpitt**
See https://cockpit-project.org/blog/tmt-cross-project-testing.html

---

That's what we just talked about in the cabal meeting. This is mostly a foundation for your planned jira card/team discussion, and a good place to collect notes, links, try notifications, etc.

I [tested this on my fork](https://github.com/martinpitt/podman/pull/1). This is [usually green](https://github.com/martinpitt/podman/runs/15525447654). I tested an [extra commit to break the checkpoint API](https://github.com/martinpitt/podman/commit/6a80963bfa11067967b7659c83678f565bb202ba) there, and it [failed the checkpoint tests](https://artifacts.dev.testing-farm.io/161f4a10-d65c-477e-ac7c-3533211b0d9c/), confirming that this really uses the podman package from the PR.

Ironically, the very first run here found an user-visible regression that isn't released yet: containers/crun#1255 .

Questions:
 - Consider not running this on rawhide (too much churn?) - can be changed at any time if/when it becomes too annoying/broken
 - Automatic notification of our team when this fails? That'd be a packit feature, it cannot efficiently be done with GitHub workflows: https://github.com/packit/packit-service/issues/1911 This isn't a blocker for the podman team, as they said they won't even look at our tests much; but it's highly desirable for the cockpit team to pick up regressions quickly

(please feel free to add more)

```release-note
None
```

### CODE REVIEW

**Comment by martinpitt** [comment]
Interesting .. I previously tested that against plain Fedora (as in my fork I don't have access to your rhcontainerbot COPR), where it was fine. Against that copr, the checkpoint test fails because [the criu binary is missing](https://artifacts.dev.testing-farm.io/440fbf32-76bb-40cc-9932-8dfbc4f1db21/work-podman-misc1vs84wdk/plans/cockpit-podman/podman-misc/execute/data/guest/default-0/test/browser/other-1/data/TestApplication-testCheckpointRestore-fedora-38-127.0.0.1-22-FAIL.png). That smells like a missing dependency somewhere, I'll have a look at that tomorrow.

**Comment by martinpitt** [comment]
On Fedora 38/39, criu is missing due to https://github.com/containers/crun/issues/1091 - apparently that bug still happens in the COPR.

On [fedora-37](https://artifacts.dev.testing-farm.io/1d90d3b7-5f9f-4bbd-9ff3-baf17d18d960/) criu is installed, but [checkpointing fails differently](https://artifacts.dev.testing-farm.io/1d90d3b7-5f9f-4bbd-9ff3-baf17d18d960/work-podman-miscrgx3vnp8/plans/cockpit-podman/podman-misc/execute/data/guest/default-0/test/browser/other-1/data/TestApplication-testCheckpointRestore-fedora-37-127.0.0.1-22-FAIL.png), with "does not support snapshot/restore". This is another real regression, so now it needs some bisecting in the COPR to find the culprit. This is a nice demonstration why I'd really like to know *right away* when a proposed change breaks stuff -- it avoids these long debugging hunts, and helps both our teams.

**Comment by martinpitt** [comment]
So that checkpoint regression is real, I filed it as https://github.com/containers/podman/issues/19506

**Comment by lsm5** [comment]
> So that checkpoint regression is real, I filed it as [containers/crun#1255](https://github.com/containers/crun/issues/1255)

should be fixed once https://github.com/containers/crun/issues/1255 merges.

**Comment by lsm5** [comment]
PR LGTM in its current form.

@martinpitt would you mind re-running these tests over a period of few days on this very PR to make the podman team kinda sorta comfortable :smile: . I did something like that for the initial packit enablement: https://github.com/containers/podman/pull/17898 . Alternatively, consider adding these to a lower traffic repo where the podman team won't mind that much.

File: .packit.yaml
Comment (lsm5) [inline .packit.yaml:22]: my bad, forgot this one. Maybe `cockpit-podman-revdeps` / `cockpit-revdeps` would make it more explicit.

Comment (lsm5) [reply .packit.yaml:22]: @martinpitt how about including the minimum set of tests that you're hoping to see in podman gating, and doing multiple runs on them? Meaning, each test run includes all of the minimum tests. As the team gets comfortable (assuming they do), we can include more.

Btw, I pinged packit folks to check for `invisible` tasks, meaning something which won't show up in the github UI for PRs, but would show up on packit's dashboard. They don't have anything like that yet and have asked me to file a feature request. Assuming that's doable, combined with `ping specific people on failure` feature request, is that something that would work for you, if the team as a whole is not comfy with including these?

Comment (martinpitt) [reply .packit.yaml:22]: > minimum set of tests that you're hoping to see in podman gating

Sorry, this is impossible to answer for me, at least in the way you hope. Let me turn the question around: Which of podman's features do you consider okay to break for users, and which not? E. g. a bare minimum test run of "I can start a container with `sleep 1` would not have caught the regression of breaking checkpoints (which would have landed in the next release).

> doing multiple runs on them

Like, re-run them 50 times to see how many times they fail? That I can do. We have a whole statistics database of test runs on our own CI, but not on the TF runs. This would be an interesting experiment.

> invisible tasks

That seems to make tracing back regressions unnecessarily hard for everyone, and is defeats the whole point of this exercise. The goal of this is to make your and my team's life more efficient and less frustrating by flagging and preventing regressions *before* they land, not sweep them under the rug (that's what we have to day). 

> "ping specific people on failure" feature

That absolutely makes sense, yes (see description). I believe it's not a hard blocker, but it would certainly make it much easier for both of us to handle problems when they happen. I'm happy to help the packit folks with implementing this. If you want to wait until that is a thing, that is certainly acceptable to me.

FTR: Of course your team has the right to say "we don't want this" or "we want to stop this experiment after one or six weeks" --  I'm just asking to give it a try. (I've gone through this discussion at least three times in my life, and after a few weeks nobody ever wanted to go back :grin: )

Comment (martinpitt) [reply .packit.yaml:22]: > > doing multiple runs on them

> Like, re-run them 50 times to see how many times they fail

Procedural warning: There is no "restart" button for passed tests, and even for failed tests I don't have the privilege to press that button anyway (as I'm not a project member). So I'll have to do that through force-pushing 50 times. That's going to create a lot of notification noise, but at least we'll have a record of all these statuses in GitHub -- if that's okay with you?

Comment (martinpitt) [reply .packit.yaml:22]: Ah, nevermind -- I can do the "50 re-runs" on my fork, like in https://github.com/martinpitt/podman/pull/1 . That avoids noise here, and also unnecessary runs of the other CI jobs (which would just be a waste of electrons).

**Comment by martinpitt** [comment]
:white_flag: I have a gut feeling that this is maybe too big a step. So I have another proposal:

 - Let's shelve this for a while, and start with cross-project testing between your own projects. These aren't affected by that packit notification feature, and don't have that smell of "we are not familiar with these".
 - Instead, let's start with sorting out your PR vs. main build COPRs, to make test results meaningful and reproducible, see https://github.com/containers/crun/pull/1256#issuecomment-1665695404  ff. I don't immediately see where the bug is, but this can't be right
 - Enable podman's Fedora gating tests in podman upstream PRs. This already tightens the feedback loop in a nice way.
 - Enable podman's tmt tests in crun and container-selinux PRs, and crun's tests in container-selinux PRs., and would have caught https://github.com/containers/crun/issues/1255 right in https://github.com/containers/crun/pull/1000 

That way, your team can become familiar with this, and land changes in these three projects with more confidence.

Then I'll come back to you after a quarter or so, and sort out packit notifications until then?

**Comment by lsm5** [comment]
> white_flag I have a gut feeling that this is maybe too big a step. So I have another proposal:
> 
> * Let's shelve this for a while, and start with cross-project testing between your own projects. These aren't affected by that packit notification feature, and don't have that smell of "we are not familiar with these".
> * Instead, let's start with sorting out your PR vs. main build COPRs, to make test results meaningful and reproducible, see [RPM: include criu dependencies crun#1256 (comment)](https://github.com/containers/crun/pull/1256#issuecomment-1665695404)  ff. I don't immediately see where the bug is, but this can't be right
> * Enable podman's Fedora gating tests in podman upstream PRs. This already tightens the feedback loop in a nice way.
> * Enable podman's tmt tests in crun and container-selinux PRs, and crun's tests in container-selinux PRs., and would have caught [podman from current rhcontainerbot/packit-builds breaks checkpointing crun#1255](https://github.com/containers/crun/issues/1255) right in [Packit: initial enablement crun#1000](https://github.com/containers/crun/pull/1000)
> 
> That way, your team can become familiar with this, and land changes in these three projects with more confidence.
> 
> Then I'll come back to you after a quarter or so, and sort out packit notifications until then?

I'm cool with that. Though I can't say for sure if a quarter will suffice given I'll be getting pushback too :smile: . I can keep you posted.

**Review by lsm5 — APPROVED** [review]
We have received approval from the team to run these tests on PRs.

/lgtm

**Comment by martinpitt** [comment]
Wohoo! You claim the prize of the first project which landed that (I have some others in the works, but they are still in review) :trophy:  . Thanks for your trust for doing this experiment! Please don't hesitate to ping me on questions or issues.

**Comment by lsm5** [comment]
> Wohoo! You claim the prize of the first project which landed that (I have some others in the works, but they are still in review) 🏆 . Thanks for your trust for doing this experiment! 

thanks for this contribution!!

> Please don't hesitate to ping me on questions or issues.

Sure thing. And I'm sure there will be some grumpy pings from our resident Gandalf @edsantiago in the near future :) . So, let's hope the packit feature to ping specific people gets done soon :D

**Comment by martinpitt** [comment]
Thanks for bearing with me! Yes, I'm sure we'll hit some bumps here and there, it's a learning experience for both of us. That TF auto-notification feature will surely be very important for scaling this up.

**Pull request merged**
